src/memoizeWith.js

```diff
@@ -0,0 +1,42 @@
+var _arity = require('./internal/_arity');
+var _curry2 = require('./internal/_curry2');
+var _has = require('./internal/_has');
+
+
+/**
+ * A customisable version of R.memoize: takes an additional function that will
+ * be applied to a given argument set and used to create the cache key under
+ * which the results of the function to be memoized will be stored. Care must
+ * be taken when implementing key generation to avoid clashes that may
+ * overwrite previous entries erroneously.
+ *
+ *
+ * @func
+ * @memberOf R
+ * @category Function
+ * @sig (*... -> a) -> (*... -> a)
+ * @param {Function} fn The function to memoize.
+ * @return {Function} Memoized version of `fn`.
+ * @see R.memoize
+ * @example
+ *
+ *      var count = 0;
+ *      var factorial = R.memoizeWith(R.head, n => {
+ *        count += 1;
+ *        return R.product(R.range(1, n + 1));
+ *      });
+ *      factorial(5); //=> 120
+ *      factorial(5); //=> 120
+ *      factorial(5); //=> 120
+ *      count; //=> 1
+ */
+module.exports = _curry2(function memoizeWith(mFn, fn) {
+  var cache = {};
+  return _arity(fn.length, function() {
+    var key = mFn(arguments);
```

Comment:
I don't think so, I would rather not as well, I just copied what you had below. The point was more that using `apply` makes the key function more friendly, eg would simply be `id` for your example function, and you can use `prop` rather than `compose(prop, head)`.